chapter.md

```diff
@@ -0,0 +1,735 @@
+# Creating Functional Forms
+
+There are many different definitions of the term "software architecture".
+Our favorite is that
+a system's architecture is what you draw on the whiteboard
+when you're explaining it to a new developer.
+Alternatively,
+the architecture of a system is a description of
+how to decompose it into pieces,
+how those pieces interact,
+and how that decomposition enables and constrains further development.
+
+In this chapter,
+we will begin our exploration of the architecture of JavaScript applications
+by looking at how to use ideas borrowed from pure functional programming
+to simplify one of the most common tasks in web development:
+form handling.
+At first glance these two topics have nothing to do with each other,
+but paradoxically,
+acting as if the state of our system couldn't be changed
+actually makes its changes easier to manage.
+
+## The Problem
+
+Users can interact with programs in a bewildering variety of ways,
+but the most popular on the web remains the HTML form.
+In a classic 1990s-style application,
+a user is presented with a form,
+fills in the fields,
+and clicks "Submit".
+This sends an HTTP POST request to a server
+with the information the user entered as its payload.
+The server validates that information,
+updates a database,
+and returns an HTML page that may or may not contain another form.
+
+As browsers have become more powerful,
+and as interaction has moved away from the desktop to mobile devices,
+developers have moved more and more of this processing cycle to the client.
+The first step was to write a little JavaScript to sanity-check the data the user entered
+before POSTing it to the server.
+Programmers quickly began using JavaScript to make form interactions more fluid
+by showing and hiding sections based on what users had entered so far,
+and by fetching bits of data to do things like auto-completion while users were still typing.
```

Comment:
somewhere between these lines it would perhaps be nice to mention that people want to validate as much as they can on the client side to avoid unnecessary requests to the server so that users don't waste their time waiting and fixing the form after the response from the backend is completed

one of the things I noticed is that both backend / front end will have the validations (backend mostly for security issues) but people want the front end smart to guide the user through and make them only do valid things =]